### PR TITLE
Feature/eslint

### PR DESCRIPTION
**Pull request by jason-fox**
- Replace jshint with eslint
- Add prettier code formatting
- Add husky and lint-staged

Incorporates and supercedes https://github.com/telefonicaid/iotagent-node-lib/pull/753

Related: https://github.com/telefonicaid/iotagent-node-lib/issues/830

### CODE REVIEW

**Comment by jason-fox** [comment]
@fgalan @AlvaroVega -

Is there anything now blocking this PR from being reviewed and merged?

**Comment by fgalan** [comment]
> @fgalan @AlvaroVega -
> 
> Is there anything now blocking this PR from being reviewed and merged?

As far as I understand by https://github.com/telefonicaid/iotagent-node-lib/issues/830#issuecomment-658635625, this PR is waiting for others to be merged first (#849 and maybe #854)

**Comment by jason-fox** [comment]
@fgalan - Given the lack of movement on #849, I guess it would make more sense to move this up the queue. #854 is dependent on #849 so would come after the others.

So I'd suggest processing in the following order - #831 , #849 then #854 

It's just that there isn't much of a queue at the moment (since the last period of activity has cleared a of of small PRs) so it would make sense to use this opportunity to get the broad ES6 change in whilst the "lull" occurs.

**Comment by fgalan** [comment]
> 
> 
> @fgalan - Given the lack of movement on #849, I guess it would make more sense to move this up the queue. #854 is dependent on #849 so would come after the others.
> 
> So I'd suggest processing in the following order - #831 , #849 then #854
> 
> It's just that there isn't much of a queue at the moment (since the last period of activity has cleared a of of small PRs) so it would make sense to use this opportunity to get the broad ES6 change in whilst the "lull" occurs.

I have updated the table at https://github.com/telefonicaid/iotagent-node-lib/issues/830#issuecomment-658222980 based on your feedback. In addition, I have asked other PRs authors about what they think about merging #831 before theirs.

If they agree (or if they don't answer in a reasonable time, assuming in that case they are missing in action :) I think we can finally progress and eventually merge PR #831

**Comment by jason-fox** [comment]
> If they agree (or if they don't answer in a reasonable time, assuming in that case they are missing in action :) I think we can finally progress and eventually merge PR #831 

@fgalan @AlvaroVega - Any chance this will be actioned before its upcoming first birthday? 🍰

File: lib/services/common/alarmManagement.js
Comment (AlvaroVega) [inline lib/services/common/alarmManagement.js:40]: are your sure about this change?

File: lib/services/commands/commandService.js
Comment (AlvaroVega) [inline lib/services/commands/commandService.js:114]: I'm not sure if this way is more easy to read....

File: lib/services/common/alarmManagement.js
Comment (jason-fox) [reply lib/services/common/alarmManagement.js:40]: EsLint and Prettier are. This is standard syntax for ES6.

File: lib/services/commands/commandService.js
Comment (jason-fox) [reply lib/services/commands/commandService.js:114]: Automated change - take it up with the guys at *Prettier*. :smile: This syntax has already been applied across the IoT Agents themselves the only hold out is the IoT-Node-lib so this is the one that isn't standardized.

**Comment by jason-fox** [comment]
@fgalan @AlvaroVega -

With the OPC-UA template change #920 , it looks like there are no objections left in the table [here](https://github.com/telefonicaid/iotagent-node-lib/issues/830#issuecomment-658222980). 
Is there anything now blocking this PR from being reviewed and merged so that  #830 can finally be closed.

**Comment by fgalan** [comment]
> @fgalan @AlvaroVega -
> 
> With the OPC-UA template change #920 , it looks like there are no objections left in the table [here](https://github.com/telefonicaid/iotagent-node-lib/issues/830#issuecomment-658222980).

I have reviewd and the only pending item is #601, but given the author seems to be MIA (Missing In Action :) as he doesn't answer to the question about adapting its PR after merging the lint changes, I think we can skip it and GO.

> Is there anything now blocking this PR from being reviewed and merged so that #830 can finally be closed.

We are near to close a new version of the iotagent-node-lib (along with some other agents). We will wait until that release (we hope that some days or a week maybe) before merging the eslint contribution (in fact, the eslint contribution could be the first PR to be merged in the new version development cycle).

**Comment by fgalan** [comment]
> We are near to close a new version of the iotagent-node-lib (along with some other agents). We will wait until that release (we hope that some days or a week maybe) before merging the eslint contribution (in fact, the eslint contribution could be the first PR to be merged in the new version development cycle).

2.14.0 has been already released (https://github.com/telefonicaid/iotagent-node-lib/releases/tag/2.14.0)

Thus, this PR would be ready to be merged, once the conflict with master gets solved.

File: lib/errors.js
Comment (fgalan) [inline lib/errors.js:10]: Typo?

File: .eslintrc.json
Comment (fgalan) [inline .eslintrc.json:9]: Comparing with the .eslintrc.json file in .eslintrc.json repo, this setting has been added here.

Which effect does it have in the eslint rules?

File: CHANGES_NEXT_RELEASE
Comment (fgalan) [inline CHANGES_NEXT_RELEASE:3]: What about?

```suggestion
Update codebase to use ES6
-  Remove JSHint and jshint overrides
-  Add esLint using standard tamia presets
-  Replace var with let/const
-  Fix or disable eslint errors
Add prettier code formatting
Add husky and lint-staged
```

To make it closer to the one used in other similar PRs, e.g. https://github.com/telefonicaid/iotagent-json/pull/456/files#diff-0bc49fac4e713c5ac9a5c346f2286f5eef1a765788273dcf4ad32b5711033e6b

File: package.json
Comment (fgalan) [inline package.json:38]: Are the new targets (prettier and prettier:text) included in the corresponding .md describing the different targets? Maybe yes, but it's a bit hard to find it in so large PR :) Could you help providing a pointer, please?

File: lib/errors.js
Comment (jason-fox) [reply lib/errors.js]: Fixed ae4416e

File: .eslintrc.json
Comment (jason-fox) [reply .eslintrc.json:9]: Same reasoning it was added [here](https://github.com/telefonicaid/iotagent-ul/pull/434/files#diff-6884918dc8291219be508e05e28965b958c734def85324f3b53858ea4702090f) - the latest version of eslint complains about Telefonica style callbacks in the code (it wants the  `callback` variable name to be unique) hence the *shadowed variable name* rule must be dropped. This is only to the updates to a more modern **Eslint**

File: package.json
Comment (jason-fox) [reply package.json:38]: Fixed ea6d399

File: CHANGES_NEXT_RELEASE
Comment (fgalan) [reply CHANGES_NEXT_RELEASE]: Fixed in 68cebf8

File: .eslintrc.json
Comment (fgalan) [reply .eslintrc.json:9]: Thanks for the clarification.

So, I understand that `"no-shadow": 0` is the desired setting and the repos not using it (if any) should be fixed. Out of the scope of this PR, of course.

NTC

**Review by fgalan — APPROVED** [review]
LGTM

It has been hard, but in the end, the PR is fine and we have found a good lull moment for it. Thanks for all your contributions!

**Comment by jason-fox** [comment]
- #937 applies ESLint to #842. 
- EsLint has also been applied to #854 and #930  which are dependent on #842 which should be reviewed/processed next.
- EsLint has also been applied to #926, which fixes one outstanding review comment

**Pull request merged**
